### ISSUE TITLE
Pulling user ID from request

### ISSUE THREAD

**Issue by ahmaidan-rr** [grafana/django-saml2-auth#52]
In my workflow I need to pull some user info from the request and then use that in the GET_METADATA_AUTO_CONF_URLS trigger.

So when I call the acs or signin views I either need to pass in the user_id that I pulled previously, or change the trigger to allow me to pass in a request. Right now these flows pass in no user_id nor the ability to pull the user_id from the trigger.

I'm new to SAML though so perhaps I'm just missing something? My project needs to support multiple organizations using the same provider, which means storing different metadata urls for different users. I'm not seeing any way of doing that with the existing signin and acs views. I don't think my use case is particularly unique though, so I'm wondering how other people are provisioning the metadata currently.

**Comment by Menda** [comment]
Hi @ahmaidan-rr I had exactly the same issue and I am preparing a PR to address it. The way I thought is that you add to your settings a function `TRIGGER.GET_USER_ID_FROM_SAML_RESPONSE`. This function living in your code should have the interface:

```
get_user_id_from_saml_response(saml_response: str, user_id: Optional[str]) -> str
```

You would get there as parameters the SAML response XML decoded and the User ID, which in your case would be `None`. What you can do inside that function is search for example for an email in the response string and that will become your `user_id` parameter in your `GET_METADATA_AUTO_CONF_URLS` trigger function.

**Comment by Menda** [comment]
https://github.com/grafana/django-saml2-auth/pull/56 @ahmaidan-rr 

Example of usage for Okta responses:

```
from lxml import etree

def get_user_id_from_saml_response(saml_response: str, user_id: Optional[str]) -> Optional[str]:
    if user_id:
        return user_id

    saml_response_bytes = str.encode(saml_response)
    tree = etree.fromstring(saml_response_bytes)
    issuer = tree.xpath(".//*[local-name()='Issuer']")
    if issuer and issuer[0].text.startswith('http://www.okta.com/'):
        name_id = tree.xpath(".//*[local-name()='NameID']")
        if name_id:
            return name_id[0].text
    return user_id
```

Add to `TRIGGER` in settings:

```
'GET_USER_ID_FROM_SAML_RESPONSE': 'get_user_id_from_saml_response',
```

**Comment by mostafa** [comment]
Thanks to @Menda, the issue is now fixed in #56. Feel free to re-open the issue if your issue isn't resolved yet.
